### PR TITLE
Issues/16

### PR DESCRIPTION
**Pull request by fjaviersanchez**
This PR tries to address #16, implementing our own derivatives of the loglikelihood to get the Fisher matrix. It is still work in progress.

### CODE REVIEW

File: augur/analyze.py [hunk 1]
```diff
@@ -17,9 +19,47 @@ def analyze(config):
...
+
+    def negloglike(x):
+        pars = deepcopy(ref_pars)  # To make sure that we start from the same point
+        for i in range(len(x)):
+            pars[par_names[i]] = x[i]
+        cosmo = firecrown.get_ccl_cosmology(pars)
+        return -ref_like(cosmo=cosmo, parameters=pars, data=data["two_point"]["data"])[0]
+
+    x0 = [data["parameters"][kk] for kk in par_names]
+    F_ij = nd.Hessian(negloglike, step=float(config["fisher"]["step"]))(x0)
```
Comment (elisachisari) [inline augur/analyze.py:64]: Ok, expression looks reasonable. How are the diff steps chosen to guarantee convergence?

File: augur/postprocess.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
+import numpy as np
+import matplotlib.pyplot as plt
+from matplotlib.collections import EllipseCollection
+import astropy.table
+from scipy.stats import norm, chi2
+import os
+
+
+def postprocess(config):
```
Comment (elisachisari) [inline augur/postprocess.py:9]: This file is not new to this branch, right? It is already in master.

**Review by elisachisari — COMMENTED** [review]
Some comments while I try to run this

File: examples/srd_y1_3x2.yaml [hunk 3]
```diff
@@ -186,25 +191,48 @@ analyze:
...
-    beta_ia:
-      kind: norm
-      loc: 1.0
-      scale: 1.6
-    eta_ia:
-      kind: norm
-      loc: 0.0
-      scale: 2.3
-    eta_ia_highz:
+    # These parameters aren't in CCL?
```
[the remark this answers is not in the document]
Comment (fjaviersanchez) [reply examples/srd_y1_3x2.yaml:194]: I don't have any strong opinions. I think they are probably better outside of CCL, but it doesn't really matter for this.

File: examples/srd_y1_3x2.yaml [hunk 4]
```diff
@@ -15,7 +15,12 @@ generate:
     h: 0.6727
     sigma8: 0.83
     n_s: 0.9645
-    # transfer_function: 'eisenstein_hu'  
+    transfer_function: "eisenstein_hu"
```
[the remark this answers is not in the document]
Comment (fjaviersanchez) [reply examples/srd_y1_3x2.yaml:18]: Oh, this is just that I changed from single to double quotes to be consistent with the rest. That's why it appears as modified.

**Comment by elisachisari** [comment]
I tried to run this new version for the Y1 yaml example and I ran into issues towards the end. I copy the log.

----------------------

Pipeline ran okay.
Applying additional prior sigma = 0.2 to params--omega_c
This will be assumed to be centered at the parameter center regardless of what the ini file says
The limits of the parameter will also not be respected.

Applying additional prior sigma = 0.8 to params--w0
This will be assumed to be centered at the parameter center regardless of what the ini file says
The limits of the parameter will also not be respected.

Applying additional prior sigma = 2.0 to params--wa
This will be assumed to be centered at the parameter center regardless of what the ini file says
The limits of the parameter will also not be respected.

Applying additional prior sigma = 0.3 to params--h
This will be assumed to be centered at the parameter center regardless of what the ini file says
The limits of the parameter will also not be respected.

Applying additional prior sigma = 0.14 to params--sigma8
This will be assumed to be centered at the parameter center regardless of what the ini file says
The limits of the parameter will also not be respected.

Applying additional prior sigma = 0.08 to params--n_s
This will be assumed to be centered at the parameter center regardless of what the ini file says
The limits of the parameter will also not be respected.

Applying additional prior sigma = 3.9 to params--a0
This will be assumed to be centered at the parameter center regardless of what the ini file says
The limits of the parameter will also not be respected.

Applying additional prior sigma = 0.5 to params--bias_lens0
This will be assumed to be centered at the parameter center regardless of what the ini file says
The limits of the parameter will also not be respected.

Applying additional prior sigma = 0.5 to params--bias_lens1
This will be assumed to be centered at the parameter center regardless of what the ini file says
The limits of the parameter will also not be respected.

Applying additional prior sigma = 0.5 to params--bias_lens2
This will be assumed to be centered at the parameter center regardless of what the ini file says
The limits of the parameter will also not be respected.

Applying additional prior sigma = 0.5 to params--bias_lens3
This will be assumed to be centered at the parameter center regardless of what the ini file says
The limits of the parameter will also not be respected.

Applying additional prior sigma = 0.5 to params--bias_lens4
This will be assumed to be centered at the parameter center regardless of what the ini file says
The limits of the parameter will also not be respected.

Postprocessing...
Generated covariance matrix was not positive definite - beware! /Users/chisa002/Documents/lsst_augur/augur/augur/postprocess.py:151: RuntimeWarning: invalid value encountered in sqrt
  heights = 2 * np.sqrt(0.5 * (C[0, 0] + C[1, 1] - D))
/Users/chisa002/Documents/lsst_augur/augur/augur/postprocess.py:161: RuntimeWarning: invalid value encountered in sqrt
  sig0 = np.sqrt(C[0, 0])
Traceback (most recent call last):
  File "/Users/chisa002/opt/anaconda3/envs/forecasting/bin/augur", line 7, in <module>
    exec(compile(f.read(), __file__, 'exec'))
  File "/Users/chisa002/Documents/lsst_augur/augur/bin/augur", line 41, in <module>
    run()
  File "/Users/chisa002/opt/anaconda3/envs/forecasting/lib/python3.9/site-packages/click/core.py", line 1137, in __call__
    return self.main(*args, **kwargs)
  File "/Users/chisa002/opt/anaconda3/envs/forecasting/lib/python3.9/site-packages/click/core.py", line 1062, in main
    rv = self.invoke(ctx)
  File "/Users/chisa002/opt/anaconda3/envs/forecasting/lib/python3.9/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/chisa002/opt/anaconda3/envs/forecasting/lib/python3.9/site-packages/click/core.py", line 763, in invoke
    return __callback(*args, **kwargs)
  File "/Users/chisa002/Documents/lsst_augur/augur/bin/augur", line 37, in run
    augur.postprocess(_config)
  File "/Users/chisa002/Documents/lsst_augur/augur/augur/postprocess.py", line 62, in postprocess
    ax[j, i].set_xlim(-sig0+centers[i], sig0+centers[i])
  File "/Users/chisa002/opt/anaconda3/envs/forecasting/lib/python3.9/site-packages/matplotlib/axes/_base.py", line 3522, in set_xlim
    left = self._validate_converted_limits(left, self.convert_xunits)
  File "/Users/chisa002/opt/anaconda3/envs/forecasting/lib/python3.9/site-packages/matplotlib/axes/_base.py", line 3439, in _validate_converted_limits
    raise ValueError("Axis limits cannot be NaN or Inf")
ValueError: Axis limits cannot be NaN or Inf

**Comment by elisachisari** [comment]
Did you compare our own forecast with one from cosmosis for the same setup?

File: augur/postprocess.py (hunk 2 above)
Comment (fjaviersanchez) [reply augur/postprocess.py:9]: No, this is in master already. I have to rebase.

File: augur/analyze.py (hunk 1 above)
Comment (fjaviersanchez) [reply augur/analyze.py:64]: I think it is a fixed step size, and at this choice just followed comosis' step size.

**Comment by fjaviersanchez** [comment]
hmmm, there's something going on with the updates. The matrix is not invertible anymore and for some reason the results don't agree 100% here is the comparison between diagonals (it might be related to the fact that the matrices are not invertible): 
image

**Comment by fjaviersanchez** [comment]
I thought it was the priors, but I added the priors in the calculation and they don't seem to make much of a difference. I am comparing here the matrices and they look alike, at least. 

`cosmosis`

image

`numdifftools`

image

**Comment by fjaviersanchez** [comment]
This PR is stale now, closing. Will resume the work on a different branch/PR

**Pull request closed without merging**